### PR TITLE
plugin/kubernetes: Enable protobuf, Update client api package

### PR DESCRIPTION
**Pull request by chrisohaver**
### 1. What does this pull request do?

* Enables protobuf in k8s api
* Update k8s.io/client-go and dependent vendored packages to latest
* Adjust code to api changes

### 2. Which issues (if any) are related?

#1094 

### 3. Which documentation changes (if any) need to be made?

none

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -5,3 +5,9 @@ coredns
 kubectl
 go-test-tmpfile*
 coverage.txt
+.idea/
+.vscode/
+cover.out
+cover.txt
+coverage.out
+myhosts
```
Comment (miekg) [inline .gitignore]: what's myhost?

**Comment by miekg** [comment]
What commit should we look at?

Can the dep stuff be separated from actual code changes? Maybe even putting a blanket 'go get' in the Makefile to pull in deps and then do the dep prune (all not ideal)

File: .gitignore (hunk 1 above)
Comment (chrisohaver) [reply .gitignore]: I didn't intend to commit .gitignore. reverting.

**Comment by chrisohaver** [comment]
I rebuilt the commit history.
code changes are in "code changes"
vendor changes are in "vendor changes"

File: plugin/kubernetes/kubernetes.go [hunk 2]
```diff
@@ -240,6 +242,7 @@ func (k *Kubernetes) initKubeCache(opts dnsControlOpts) (err error) {
 	if err != nil {
 		return err
 	}
+	config.ContentType = "application/vnd.kubernetes.protobuf"
```
Comment (miekg) [inline plugin/kubernetes/kubernetes.go]: Why don't we set this in k.getClientConfig?

File: plugin/kubernetes/handler_test.go [hunk 3]
```diff
@@ -206,10 +207,10 @@ func (APIConnServeTest) PodIndex(string) []interface{} {
 	return a
 }
 
-func (APIConnServeTest) ServiceList() []*api.Service {
+func (APIConnServeTest) ServiceList() []interface{} {
```
Comment (miekg) [inline plugin/kubernetes/handler_test.go]: Why does this need to be moved to empty interface slice? We loose all type safety

**Review by miekg — CHANGES_REQUESTED** [review]
Thanks for splitting  into 2 comments.

**Comment by miekg** [comment]
Thanks, can you squash  the code changes again?

File: plugin/kubernetes/controller.go [hunk 4]
```diff
@@ -341,62 +218,59 @@ func (dns *dnsControl) Stop() error {
...
 		return nil
 	}
-	pods, err := dns.podLister.Indexer.ByIndex(podIPIndex, ip)
+	os, err := dns.podLister.ByIndex(podIPIndex, ip)
 	if err != nil {
 		return nil
 	}
-
+	for _, o := range os {
+		p, ok := o.(*api.Pod)
```
Comment (miekg) [inline plugin/kubernetes/controller.go:249]: again, is this really needed? Or can this API really return something else?

File: plugin/kubernetes/controller.go [hunk 5]
```diff
@@ -341,62 +218,59 @@ func (dns *dnsControl) Stop() error {
...
-}
-
-func (dns *dnsControl) ServiceList() []*api.Service {
-	svcs, err := dns.svcLister.List(labels.Everything())
-	if err != nil {
-		return []*api.Service{}
+func (dns *dnsControl) ServiceList() (svcs []*api.Service) {
+	os := dns.svcLister.List()
+	for _, o := range os {
+		s, ok := o.(*api.Service)
```
Comment (miekg) [inline plugin/kubernetes/controller.go:231]: Are these checks really needed?

**Review by miekg — COMMENTED** [review]
I see what had these checks before. I always wondered, are they needed? It's an extra loop for a k8s queries, so if then can be dropped, it would matter a tiny bit.

File: plugin/kubernetes/controller.go (hunk 4 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go:249]: I dont know. I was coding defensively...  Should I remove the checks?

Comment (miekg) [reply plugin/kubernetes/controller.go:249]: dunno... @johnbelamaric any ideas?

Comment (miekg) [reply plugin/kubernetes/controller.go:249]: we can remove them later.

**Review by miekg — APPROVED** [review]

File: plugin/kubernetes/controller.go (hunk 4 above)
Comment (chrisohaver) [reply plugin/kubernetes/controller.go:249]: > It's an extra loop for a k8s queries..

The loop is required to convert the types to the return type.  This is why I was doing the conversion outside the function originally.

> Are these checks really needed

This i dont know for sure ... But i think that if they always returned one type, the methods would be typed that way, hence the check.

Comment (johnbelamaric) [reply plugin/kubernetes/controller.go:249]: I wonder how the well type assertions are optimized by the compiler. I would imagine the overhead would be really small.

Comment (johnbelamaric) [reply plugin/kubernetes/controller.go:249]: Also none of this stuff is in-band.

Comment (johnbelamaric) [reply plugin/kubernetes/controller.go:249]: I mean, none of this runs for a DNS query, it's all async stuff in other go routines.

**Comment by johnbelamaric** [comment]
Oh never mind, @chrisohaver has corrected me, this is in-band. I was thinking this was in a different part of the code.

Nonetheless, I do think the overhead is likely very low for the type assertions. Would be interesting to check.

**Pull request merged**